Add keygen utility

dberliner opened the pull request:
This code currently includes code to login with XCR, generate an EC key and use it to sign a X509 certificate.

drbild commented on keygen/src/json_model.h:
```diff
@@ -0,0 +1,52 @@
+
+#include <jansson.h>
+
+struct xcr_login {
```
It might be helpful for internal consistency to mimic the naming scheme from the [Go SDK](https://github.com/xaptum/go-enf/blob/master/enf/auth.go).

So 
- `xcr.{h,c}` would become `auth.{h,c}`.
- `xcr_login` would become `auth_request`
- `xcr_auth_data` would become `credentials`
- `xcr_page` would become `auth_page`.


The `xiam` functions aren't in the `go-enf` API yet (or in `enfun-keygen`!) , but those would get names like
- `iam.{h.c}`
- `iam_create_endpoint_request` (a `NEW_ENDPOINT_AUTH` object)
- `iam_endpoint` (the returned `ENDPOINT` object)

drbild commented on keygen/src/xcr_comm.c:
```diff
@@ -0,0 +1,149 @@
+ */
+struct xcr_auth *xcr_send_login(struct xcr_session *session, struct xcr_login *credentials)
+{
+    json_t *xcr_proc_auth_resp = NULL;
```
Wow, the cURL API is messy!  Lots of code for what would be one line in Python or another memory-managed, high-level language!

I expect that you'll be able to refactor a bunch of this into smaller utility methods, to make long term maintenance easier.  But probably best to wait until after/when adding the IAM API, to really see what the commonalities are.

drbild commented on keygen/src/xcr_comm.h:
```diff
@@ -0,0 +1,35 @@
+ * limitations under the License.
+ */
+
+#ifndef _XCR_COMM_H
```
I personally would combine this with `json_model.h` as `auth.{h,c}`.  But its fine to separate them if you prefer: `auth_model.{hcc}` and `auth_api.{h,c}`.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+void save_EC_private(EVP_PKEY *pkey, char *fname)
+{
+    FILE *f = fopen(fname, "wb");
+    PEM_write_PrivateKey(f, pkey, NULL,NULL,0,NULL,NULL);
```
I'm not sure how you're planning on calling `save_EC_privatekey`, but the output of the `gen_EC` (a `EC_KEY` structure) has to be converted into a `PKEY` structure before calling `PEM_write_PrivateKey`.

So I think you'll need code like
```
    EVP_PKEY *pkey = EVP_PKEY_new();
    if (!pkey) {
        <handle error>
    }   

    if (!EVP_PKEY_assign_EC_KEY(pkey, ec_key)) {
        <handle error>
    }   
```

It looks like OpenSSL does have a ecc-specific write function, but it's only v1.1 (so, now that we're supporting v1.0, it's probably easier to just convert to a PKEY no matter the version)

dberliner replied:
I had initially thought the caller would need both the key and pkey, but that doesn't appear to be the case so I'm just going to be passing an `EVP_PKEY *` around. The docs say that when the pkey is free'd it will also free the underlying key.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+ */
+X509 *gen_cert(EVP_PKEY *pkey, char *cn)
+{
+    X509 *x509 = X509_new();
```
As with the `EC_KEY`, how does this get `free`'d later?

dberliner replied:
I wasn't managing that. I added a destroy function but it's a simple call to `X509_free`

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+
+#include <stdio.h>
+#include <string.h>
+#include <openssl/evp.h>
```
I think some of these OpenSSL headers aren't required (definitely not the `rsa` one). You probably only need `ec.h`, `pem.h`, and `x509.h`.

dberliner replied:
Trimmed.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+    X509_NAME * name = X509_get_subject_name(x509);
+
+    /* Set the country code and common name. */
+    X509_NAME_add_entry_by_txt(name, "C",  MBSTRING_ASC, (unsigned char *)"US",        -1, -1, 0);
+    X509_NAME_add_entry_by_txt(name, "O",  MBSTRING_ASC, (unsigned char *)"Xaptum", -1, -1, 0);
```
I don't know that we need to set the Company or Country (these certs belong to someone else, and they might be in a different country, anyhow)

dberliner replied:
Removed

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+    /* Set the country code and common name. */
+    X509_NAME_add_entry_by_txt(name, "C",  MBSTRING_ASC, (unsigned char *)"US",        -1, -1, 0);
+    X509_NAME_add_entry_by_txt(name, "O",  MBSTRING_ASC, (unsigned char *)"Xaptum", -1, -1, 0);
+    X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)cn, -1, -1, 0);
```
The return value needs to be checked here.

dberliner replied:
Checked

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+
+    /* Now set the issuer name. */
+    X509_set_issuer_name(x509, name);
+
```
We need to actually sign the cert!

You need something like
```
    if (!X509_sign(x509, pkey, EVP_sha256())) {
        <handle error>
    }
```

dberliner replied:
I was having the caller do the signing, but there's no reason it can't be done in the function, changed.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+void write_cert(X509 *cert, char *fname)
+{
+    FILE *f;
+    f = fopen(fname, "wb");
```
Again, this shouldn't include `b`.

dberliner replied:
Removed

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+        goto out;
+    }
+
+    /* Set the serial number. */
```
Sorry, I forgot to include this in my actual review.

We should actually generate a random serial number (both to help us track these certs, and also as general practice it looks bad to have all our certs have the same serial number).

So, we need something like
```
    unsigned char buffer[19];    // to hold a random number up to 2**152
    randombytes_buf(buffer, sizeof(buffer));
    BIGNUM *bn = BN_bin2bn(buffer, sizeof(buffer), NULL);
    if (!bn) {
        <handle error>
    }   

    ASN1_INTEGER *serial = BN_to_ASN1_INTEGER(bn, NULL);
    if (!serial) {
        <handle error>
    } 

    if (!X509_set_serialNumber(x509, serial)) {
        <handle error>
    }
```

This requires that we `#include <sodium.h>`. This now uses the `libsodium` library for generating the random serial number; that should be fine, since `enftun` already has to link against libsodium for `xtt`.

dberliner replied:
Done

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+
+    if (ec_key && !EC_KEY_generate_key(ec_key))
+    {
+        EC_KEY_free(ec_key);
```
The key is used to sign the cert so it needs to exist outside of the function.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+ */
+void save_EC_private(EVP_PKEY *pkey, char *fname)
+{
+    FILE *f = fopen(fname, "wb");
```
Check added

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+ */
+void save_EC_private(EVP_PKEY *pkey, char *fname)
+{
+    FILE *f = fopen(fname, "wb");
```
Removed

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+void save_EC_private(EVP_PKEY *pkey, char *fname)
+{
+    FILE *f = fopen(fname, "wb");
+    PEM_write_PrivateKey(f, pkey, NULL,NULL,0,NULL,NULL);
```
Check added

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,117 @@
+void save_EC_private(EVP_PKEY *pkey, char *fname)
+{
+    FILE *f = fopen(fname, "wb");
+    PEM_write_PrivateKey(f, pkey, NULL,NULL,0,NULL,NULL);
```
Changed

zanebeckwith commented:
@dberliner I'm still unclear how the key- and cert-generation functions are intended to be used. I'm wary of creating a too-complex API for users, in case they misuse it (leading to memory leaks and potentially to sensitive information exposure).

Is there any need for the caller to receive any of the OpenSSL objects (e.g. the PKEY or the X509)? Or, is the intended usage of these functions just to create a keypair and a self-signed cert, which are written to disk?

If the latter, I'd advocate we just expose a *single* function, to do exactly that. It would take only the IPv6 Common Name and the filenames for the key and cert files, and return success or failure codes (and the correctly-serialized public key, I guess, so you can then POST to the XIAM).

dberliner commented:
> @dberliner I'm still unclear how the key- and cert-generation functions are intended to be used. I'm wary of creating a too-complex API for users, in case they misuse it (leading to memory leaks and potentially to sensitive information exposure).
> 
> Is there any need for the caller to receive any of the OpenSSL objects (e.g. the PKEY or the X509)? Or, is the intended usage of these functions just to create a keypair and a self-signed cert, which are written to disk?
> 
> If the latter, I'd advocate we just expose a _single_ function, to do exactly that. It would take only the IPv6 Common Name and the filenames for the key and cert files, and return success or failure codes (and the correctly-serialized public key, I guess, so you can then POST to the XIAM).

I should have been clear about this in the description. What you're describing is going to be the final product. Once the XIAM and enftun config code is completed I am going to write a caller that handles this code and users will have a command line application.

zanebeckwith commented on keygen/src/main.c:
```diff
@@ -0,0 +1,231 @@
+    }
+    printf("Successful.\n");
+
+    /* Seed srand for the encryption functions */
+    srand(time(NULL));
```
Related to this, though: you do need to initialize the OpenSSL and libsodium libraries early in this `main` function.

I believe all you need for init'ing OpenSSL is
```
SSL_library_init();
```

That should work for either `v1.0` or `v1.1`

For initializing libsodium, here's the recommended cruft:
```
    int init_ret;
#if defined(__linux__) && defined(RNDGETENTCNT)
    int rand_fd;
    int ent_count;

    if ((rand_fd = open("/dev/random", O_RDONLY)) != -1) {
        if (ioctl(rand_fd, RNDGETENTCNT, &ent_count) == 0 && ent_count < 160) {
            return XTT_RETURN_INSUFFICIENT_ENTROPY;
        }

        (void) close(rand_fd);
    }   
#endif

    init_ret = sodium_init();
    if (init_ret == -1) {
        return XTT_RETURN_BAD_INIT;
    } else {    /* Can also include init_ret == 1, indicating libsodium already initialized */
        return 0;
    }   
```

dberliner replied:
I'm using this code now. Tested on both versions of openssl.

dberliner replied:
For `srand` it was unnecessary so removed. I read somewhere early on that it should be seeded and didn't give it much thought afterwards.

zanebeckwith replied:
Ha, whenever you read anything about using a crypto library, and they mention you should do anything with the stdlib rand, ignore everything they say. The stdlib rand is not secure at all for cryptographic usage. That was my alarm when I saw this.

zanebeckwith commented on keygen/src/main.c:
```diff
@@ -0,0 +1,231 @@
+    /* Clean up */
+    destroy_EC_PKEY(pkey);
+    ep_auth_resp_destroy(&ep_resp);
+
```
`cert` needs to be `free`'d, too

dberliner replied:
Made irrelevant due to other changes. Cert doesn't leave `write_cert` now.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    /* Set the serial number. */
+    ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);
```
Removed

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+        ec_key = NULL;
+    }
+
+    EVP_PKEY_set1_EC_KEY(pkey, ec_key);
```
Good catch, I hadn't given any though to version differences. Fixed (it did produce an error when I downgraded)

dberliner replied:
I've gone over pretty much the entire program cleaning up the allocation and free processes. Pretty much anything that can be a stack variable is now.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+ */
+char *public_key_encode_x962(EVP_PKEY *pkey)
+{
+    EC_KEY *key = EVP_PKEY_get0_EC_KEY(pkey);
```
Good catch, I hadn't given any though to version differences. Fixed (it did produce an error when I downgraded)

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+    char *hex;
+
+    hex = EC_POINT_point2hex(EC_KEY_get0_group(key), point, EC_KEY_get_conv_form(key), NULL);
+    printf("Encoded key: %s\n", hex);
```
Correct. Removed.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+    const EC_POINT * point = EC_KEY_get0_public_key(key);
+    char *hex;
+
+    hex = EC_POINT_point2hex(EC_KEY_get0_group(key), point, EC_KEY_get_conv_form(key), NULL);
```
Fixed.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+    /* We want to copy the subject name to the issuer name. */
+    X509_NAME * name = X509_get_subject_name(x509);
+
+    /* Set the country code and common name. */
```
Fixed.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+        goto err_freeserial;
+    }
+    /* Success! */
+    goto out;
```
Fixed.

dberliner replied:
Agreed, pulling serial out to its own function makes sense here.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+{
+    unsigned char buffer[19];    // to hold a random number up to 2**152
+    randombytes_buf(buffer, sizeof(buffer));
+    return BN_bin2bn(buffer, sizeof(buffer), NULL);;
```
Agreed, done.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,243 @@
+{
+    unsigned char buffer[19];    // to hold a random number up to 2**152
+    randombytes_buf(buffer, sizeof(buffer));
+    return BN_bin2bn(buffer, sizeof(buffer), NULL);;
```
Removed

dberliner commented:
@zanebeckwith I've implemented all of your comments, let me know if you have any other concerns.
@drbild I've implemented Zanes commends and done a few of them across the whole program. The memory allocation and freeing is a lot more consistent now.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+        goto out;
+    }
+
+    hex = EC_POINT_point2hex(EC_KEY_get0_group(key), point, EC_KEY_get_conv_form(key), NULL);
```
Rather than use EC_KEY_get_conv_form, we should specify `POINT_CONVERSION_UNCOMPRESSED`.

I think the `xiam` documentation isn't clear on it, but I think somewhere (maybe in the XMB) we assume uncompressed x9.62 format.

dberliner replied:
Changed.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+    if (!ec_key) {
+        fprintf(stderr, "%s ERR: Could not generate EC_KEY", __func__);
+        ret = 0;
+        goto out;
```
`ret` will be returned unset here. Maybe just initialize `ret` to `0`?

dberliner replied:
Ret is set to 0 here, nothing wrong with initializing it though.

zanebeckwith replied:
Ha, that's weird, sorry, don't know what I was looking at when I wrote that comment

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+{
+    if(key_in->pkey)
+        EVP_PKEY_free(key_in->pkey);
+    OPENSSL_free(key_in->hex_str);
```
Need to check `hex_str` for `NULL`, too, right?

dberliner replied:
`OPENSSL_free` is like `free` with NULL input. The documentation for `EVP_PKEY_free` specifies the same so I've removed that NULL check.

zanebeckwith replied:
I always forget that `free`'ing a `NULL` pointer is fine!

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+
+    /* Success */
+    ret = 1;
+    goto out;
```
`ec_key` isn't `free`'d on success.

dberliner replied:
It shouldn't be. From the docs for `EVP_PKEY_assign_EC_KEY`
```EVP_PKEY_assign_EC_KEY() also set the referenced key to key however these use the supplied key internally and so key will be freed when the parent pkey is freed.```

I've removed the `ec_key` free from and added a separate cleanup section for when `ec_key` is allocated but the assignment fails.

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+ *
+ * Return: A new buffer containing the x9.62-encoded string
+ */
+static char *public_key_encode_x962(EVP_PKEY *pkey)
```
Rather than take the `EVP_PKEY` for this, can't we just take the `EC_KEY` (`gen_key` still has the `ec_key` when it calls this function). That avoids having to create and free the `EC_KEY` here

dberliner replied:
Agreed, changed.

dberliner commented on keygen/src/main.c:
```diff
@@ -0,0 +1,309 @@
+    if (!ret) {
+        printf("OpenSSL init failed. Exiting.\n");
+        goto cleanup;
+    }
```
Changed.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+    key = EVP_PKEY_get1_EC_KEY(pkey);
+    if (!key) {
+        printf("%s Could not get EC_KEY from pkey", __func__);
+        goto out;
```
No longer relevant.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+ * Caller must free the returned BIGNUM
+ *
+ * Return: A new random BIGNUM
+ */
```
Changed

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+    bignum = BN_bin2bn(buffer, sizeof(buffer), NULL);
+
+    if (!bignum) {
+        ret = 0;
```
Yes, this was a vestige from the functions old approach.

dberliner commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,327 @@
+
+    /* This certificate is valid from now until exactly 365 days from now */
+    X509_gmtime_adj(X509_get_notBefore(x509), 0);
+    X509_gmtime_adj(X509_get_notAfter(x509), 31536000L);
```
That works fine.

dberliner commented:
All comments have been addressed and a few others fixed with Valgrind. Valgrind isn't showing any obvious leaks (everything I see is internal curl related and expected).

zanebeckwith commented on keygen/src/ssl.c:
```diff
@@ -0,0 +1,325 @@
+
+    if (!bignum) {
+        ret = NULL;
+        fprintf(stderr, "%s ERR: Could generate serial number", __func__);
```
Missing "not" in the error message

zanebeckwith commented on keygen/CMakeLists.txt:
```diff
@@ -0,0 +1,62 @@
+find_package(OpenSSL REQUIRED QUIET 1.1)
+find_package(Jansson REQUIRED)
+find_package(CURL REQUIRED)
+if (NOT TARGET sodium)
```
libsodium is used for generating the random serial number for the x.509 certs. I figured just initializing the libsodium library and calling their random_bytes_buf function is easier than dealing with seeding the OpenSSL rng, since we already require that library, as you said.

zanebeckwith replied:
I was wrong, the OpenSSL rng does automatically seed itself, so we could use that instead and not have to fiddle with seeding it (at least, I feel comfortable with the corner cases for this use-case).

It isn't critical to use cryptographically-strong random bytes for the serial number (this is a self-signed cert, so any attacker can do whatever they want outside of this project). I just didn't want to use the same value for every cert, or use `rand`, or have to fiddle with seeding.

It doesn't seed itself if `/dev/urandom` isn't available (and, apparently, it can use CryptGenRandom on Windows), but that's probably the same limitation as libsodium. I don't know if it fails silently or what, but again this isn't that sensitive so it's probably OK.

You've gotta love OpenSSL documentation: https://wiki.openssl.org/index.php/Random_Numbers#Seeds
"Even if the docs state a seed is not needed, you should seed it anyway."
... a few lines later ...
"OpenSSL will attempt to seed the random number generator automatically upon instantiation"
Also, "Be careful when deferring to RAND_poll on some Unix systems because it does not seed the generator". So, we'll automatically seed it for, except for when we won't.
And "(the seeding can have) negative interactions on newer Windows platforms, so your program could hang or crash"

I don't know if we need to be concerned about this for our other use of OpenSSL (actually generating the keys and signatures, running the TLS session in the main library, etc.). Probably. Ugh.

drbild commented on keygen/src/auth.h:
```diff
@@ -0,0 +1,60 @@
+
+#include <jansson.h>
+
+struct xcr_session {
```
`xcr_session` is not used and can be removed.

drbild commented on keygen/src/auth.h:
```diff
@@ -0,0 +1,60 @@
+ * limitations under the License.
+ */
+
+#ifndef _XCR_COMM_H
```
These `#ifdefs` should how be `_AUTH_H`

drbild commented on keygen/src/auth.h:
```diff
@@ -0,0 +1,60 @@
+    char token[512];
+};
+
+struct xcr_login {
```
Maybe `auth_login` instead of `xcr_login`, so that the prefix is consistent within this file?  Ditto for the functions below:  `xcr_send_login` -> `auth_login`, etc.  The

drbild commented on keygen/src/curl.h:
```diff
@@ -0,0 +1,27 @@
+//
```
Remove this header

drbild commented on keygen/src/main.c:
```diff
@@ -0,0 +1,309 @@
+{
+    const char *help_msg =
+        "Usage: %s -a NETWORK -c CONF_FILE -u API_EMAIL\n"
+        "NETWORK: The IPv6 subnet you are provisioning in. Eg. 2607:8F80:DEAD:BEEF::/64\n"
```
A few suggestions for the help message.

We need to support both a network  (::/64) and a full address.  It would also be nice to provide a bit more explanation.

 So how about:
```
const char *help_msg =
    "Provision and register credentials (key and cert) for enftun to connect.\n"
    "\n"
    "Usage: %s -c CONF_FILE -a ADDRESS -u USERNAME\n"
    "CONF_FILE: Path to the config file for the enftun instance (/etc/enftun/enf0.conf)\n"
    "ADDRESS: The IPv6 address to provision credentials for. If a ::/64 network is provided, a new, randomly generated IPv6 address in the network will created.\n"
    "USERNAME: Your ENF API username.\n"
    "\n"
    "This program will prompt for your password.\n"
```

drbild commented:
@dberliner Also need to get the TravisCI builds passing.   I think the failures were just due to missing dependencies (curl, jansson) in the Docker container.

I've pushed version 0.1.2 of the docker container. Please update the `.travis.yml` file to use that version.  Replace `enftun-build:0.1.1` with `enftun-build:0.1.2` on line 33.

dberliner commented:
@drbild 
1. Fixed. The API expects the field to be explicitly named. Some changed required to the IAM structure. `union endpoint` became `char endpoint[44];` and a new enum is used to determine if it is an address or network. The creation function looks for `/64` to decide the two. Based on that the respective API fields are populated. 
2. Changed. An optionional `-e` parameter was added defaulting to `/etc/enftun` that will only be paid attention to if the path is relative.
3. Changed.
